x/crosschain/keeper/msg_server_whitelist_erc20.go

```diff
@@ -0,0 +1,122 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"github.com/zeta-chain/zetacore/common"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	fungibletypes "github.com/zeta-chain/zetacore/x/fungible/types"
+	zetaObserverTypes "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func (k Keeper) WhitelistERC20(goCtx context.Context, msg *types.MsgWhitelistERC20) (*types.MsgWhitelistERC20Response, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if msg.Creator != k.zetaObserverKeeper.GetParams(ctx).GetAdminPolicyAccount(zetaObserverTypes.Policy_Type_deploy_fungible_coin) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "Deploy can only be executed by the correct policy account")
+	}
+	erc20Addr := ethcommon.HexToAddress(msg.Erc20Address)
+	if erc20Addr == (ethcommon.Address{}) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid ERC20 contract address (%s)", msg.Erc20Address)
+	}
+
+	// check if it's already whitelisted
+	foreignCoins := k.fungibleKeeper.GetAllForeignCoins(ctx)
+	for _, fcoin := range foreignCoins {
+		assetAddr := ethcommon.HexToAddress(fcoin.Asset)
+		if assetAddr == erc20Addr && fcoin.ForeignChainId == msg.ChainId {
+			return nil, sdkerrors.Wrapf(types.ErrInvalidAddress, "ERC20 contract address (%s) already whitelisted on chain (%d)", msg.Erc20Address, msg.ChainId)
+		}
+	}
+	// TODO: check if this chain is supported
+	chain := common.GetChainFromChainID(msg.ChainId)
+
+	tmpCtx, commit := ctx.CacheContext()
+	// add to the foreign coins. Deploy ZRC20 contract for it.
+	zrc20Addr, err := k.fungibleKeeper.DeployZRC20Contract(tmpCtx, msg.Name, msg.Symbol, uint8(msg.Decimals), chain.ChainId, common.CoinType_ERC20, msg.Erc20Address, big.NewInt(msg.GasLimit))
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrDeployContract, "failed to deploy ZRC20 contract for ERC20 contract address (%s) on chain (%d)", msg.Erc20Address, msg.ChainId)
+	}
+	if zrc20Addr == (ethcommon.Address{}) {
+		return nil, sdkerrors.Wrapf(types.ErrDeployContract, "deployed ZRC20 return 0 address for ERC20 contract address (%s) on chain (%d)", msg.Erc20Address, msg.ChainId)
+	}
+	// add to the foreign coins
+	foreignCoin := fungibletypes.ForeignCoins{
+		Zrc20ContractAddress: zrc20Addr.Hex(),
+		Asset:                msg.Erc20Address,
+		ForeignChainId:       msg.ChainId,
+		Decimals:             msg.Decimals,
+		Name:                 msg.Name,
+		Symbol:               msg.Symbol,
+		CoinType:             common.CoinType_ERC20,
+		GasLimit:             uint64(msg.GasLimit),
+	}
```

Comment:
`foreignCoin` is only used at the end of the method with:
```
k.fungibleKeeper.SetForeignCoins(ctx, foreignCoin)
```
I think the code would be more visual by putting this at the end
```
// add to the foreign coins
foreignCoin := fungibletypes.ForeignCoins{
	Zrc20ContractAddress: zrc20Addr.Hex(),
	Asset:                msg.Erc20Address,
	ForeignChainId:       msg.ChainId,
	Decimals:             msg.Decimals,
	Name:                 msg.Name,
	Symbol:               msg.Symbol,
	CoinType:             common.CoinType_ERC20,
	GasLimit:             uint64(msg.GasLimit),
}
k.fungibleKeeper.SetForeignCoins(ctx, foreignCoin)
```